tests/bsim/bluetooth/host/att/long_read/testlib/att_read.c

```diff
@@ -0,0 +1,390 @@
+			.start_handle = start_handle,
+			.end_handle = end_handle,
+			.func = gatt_discover_service_cb,
+			.uuid = uuid,
+		}};
+	struct bt_testlib_gatt_discover_service_closure *const ctx = &_ctx;
+
+	k_mutex_init(&ctx->lock);
+	k_condvar_init(&ctx->done);
+
+	__ASSERT_NO_MSG(conn);
+	__ASSERT_NO_MSG(IN_RANGE(start_handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE,
+				 BT_ATT_LAST_ATTRIBUTE_HANDLE));
+	__ASSERT_NO_MSG(
+		IN_RANGE(end_handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE));
+
+	k_mutex_lock(&ctx->lock, K_FOREVER);
+
+	api_err = bt_gatt_discover(conn, &ctx->params);
+
+	if (!api_err) {
+		k_condvar_wait(&ctx->done, &ctx->lock, K_FOREVER);
+	}
+
+	k_mutex_unlock(&ctx->lock);
+
+	if (api_err) {
+		__ASSERT_NO_MSG(api_err < 0);
+		return api_err;
+	}
+	__ASSERT_NO_MSG(ctx->att_err >= 0);
+	return ctx->att_err;
+}
+
+struct bt_testlib_gatt_discover_char_closure {
+	struct bt_gatt_discover_params params;
+	uint8_t att_err;
+	uint16_t *const result_def_handle;
+	uint16_t *const result_value_handle;
+	uint16_t *const result_end_handle;
+	uint16_t svc_end_handle;
+	struct k_mutex lock;
+	struct k_condvar done;
+};
+
+static inline uint8_t gatt_discover_char_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+					    struct bt_gatt_discover_params *params)
+{
+	LOG_DBG("gatt_discover_char_cb conn %u attr %p params %p", bt_conn_index(conn), attr,
+		params);
+
+	struct bt_testlib_gatt_discover_char_closure *ctx =
+		CONTAINER_OF(params, struct bt_testlib_gatt_discover_char_closure, params);
+	bool read_more = false;
+
+	k_mutex_lock(&ctx->lock, K_FOREVER);
+
+	if (ctx->att_err == BT_ATT_ERR_ATTRIBUTE_NOT_FOUND) {
+		/* The start of the charachteristic was not found yet.
+		 * This is the start of the characteristic.
+		 */
+		if (attr) {
+			ctx->att_err = BT_ATT_ERR_SUCCESS;
+			if (ctx->result_def_handle) {
+				*ctx->result_def_handle = attr->handle;
+			}
+			if (ctx->result_value_handle) {
+				*ctx->result_value_handle =
+					((struct bt_gatt_chrc *)attr->user_data)->value_handle;
+			}
+			if (ctx->result_end_handle) {
+				read_more = true;
+			}
+		}
+	} else {
+		/* This is the end of the characteristic.
+		 */
+		if (attr) {
+			__ASSERT_NO_MSG(ctx->result_end_handle);
+			*ctx->result_end_handle = (attr->handle - 1);
+		}
+	};
+
+	if (!read_more) {
+		k_condvar_signal(&ctx->done);
+	}
+	k_mutex_unlock(&ctx->lock);
+	return read_more ? BT_GATT_ITER_CONTINUE : BT_GATT_ITER_STOP;
+}
+
+int bt_testlib_gatt_discover_characteristic(uint16_t *const result_value_handle,
+					    uint16_t *const result_end_handle,
+					    uint16_t *const result_def_handle, struct bt_conn *conn,
+					    const struct bt_uuid *uuid, uint16_t start_handle,
+					    uint16_t svc_end_handle)
+{
+	int api_err;
+
+	if (result_end_handle) {
+		/* If there is no second result, the end_handle is the svc_end. */
+		*result_end_handle = svc_end_handle;
+	}
+
+	struct bt_testlib_gatt_discover_char_closure _ctx = {
```

Comment:
Do you mean to ask why `ctx` is a pointer? Or why `ctx` is not initialized using an object literal?

Reply:
Both, or rather why we need the pointer in the first place? 

This is also the first and only place in the entire Zephyr codebase where we have declared a variable starting with the `_` prefix, so that's definitely irregular.

Reply:
I guess it's fine. Not a fan of the previously unused `_` prefix for such variables (especially since the same prefix is used for internal fields in public structs), but I guess it's not harmful. 

Could potentially rename to something like `ctx_val`, but I guess it is also fine to keep as is